File: week1/alfazya021/15-3Sum.js
var threeSum = function (a) {
  let answer = [];
  let nums = a;

  nums.sort(function (a, b) {
    return a - b;
  });

  for (let i = 0; i < nums.length - 2; i++) {
    if (i == 0 || (i > 0 && nums[i] != nums[i - 1])) {
      let low = i + 1;
      let high = nums.length - 1;
      let sum = 0 - nums[i];

      while (low < high) {
        let x = [nums[i], nums[low], nums[high]];

        if (nums[low] + nums[high] === sum) {
          answer.push(x);

          while (low < high && nums[low] == nums[low + 1]) low++;
          while (low < high && nums[high] == nums[high - 1]) high--;

          low++;
          high--;
        } else if (nums[low] + nums[high] > sum) {
          high--;
        } else {
          low++;
        }
      }
    }
  }

  return answer;
};

File: week1/atakantekatan/15. 3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = []
        for i in range(len(nums)-2):
            if i>0 and nums[i] == nums[i-1]:
                continue
            left = i+1
            right = len(nums)-1
            if nums[i]>0:
                break
            while left<right:
                SUM = nums[i] + nums[left] + nums[right]
                if SUM < 0:
                    left += 1
                elif SUM >0:
                    right -= 1
                else:
                    LIST = [nums[i],nums[left],nums[right]]
                    ans.append(LIST)
                    while left<right and nums[left]==LIST[1]:
                        left += 1
                    while left<right and nums[right]==LIST[2]:
                        right -= 1
        return ans